Welcome to...
-------------

.. code::

    █████╗  ███╗   ██╗ ███╗   ██╗  ██████╗  ███████╗ ███╗   ██╗  █████╗  ██╗  ██╗ ███████╗
   ██╔══██╗ ████╗  ██║ ████╗  ██║ ██╔═══██╗ ██╔════╝ ████╗  ██║ ██╔══██╗ ██║ ██╔╝ ██╔════╝
   ███████║ ██╔██╗ ██║ ██╔██╗ ██║ ██║   ██║ ███████╗ ██╔██╗ ██║ ███████║ █████╔╝  █████╗  
   ██╔══██║ ██║╚██╗██║ ██║╚██╗██║ ██║   ██║ ╚════██║ ██║╚██╗██║ ██╔══██║ ██╔═██╗  ██╔══╝  
   ██║  ██║ ██║ ╚████║ ██║ ╚████║ ╚██████╔╝ ███████║ ██║ ╚████║ ██║  ██║ ██║  ██╗ ███████╗
   ╚═╝  ╚═╝ ╚═╝  ╚═══╝ ╚═╝  ╚═══╝  ╚═════╝  ╚══════╝ ╚═╝  ╚═══╝ ╚═╝  ╚═╝ ╚═╝  ╚═╝ ╚══════╝


|

...a Snakemake workflow streamlining taxonomic and functional annotation of metagenomes and metagenome-assembled genomes (MAGs). It takes raw sequencing reads, assembles and annotates contigs, and outputs colorful graphics and rich tables. You can choose between different types of input read formats (paired-end or interleaved), annotation databases (eg. Pfam, KEGG, CAZy), and more. 

|
.. admonition:: If in doubt, reach out!

   If you have any questions or issues to report, please go to `GitHub <https://github.com/bheimbu/annoSnake>`_ and open a :guilabel:`New issue`.

|

Content
-------
.. toctree::
   for_the_impatient
   installation
   step_by_step
   databases
   output_overview
   workflow_overview
   
