.. _output::
Output
======

.. contents::
   :local:
   :backlinks: none

Tables
^^^^^^

Tables are given in **CSV** format and are ready-to-use for downstream analysis. They can be found in the :file:`{{OUTDIR}}/tables` directory.

|

.. csv-table:: **Example table:** Relative abundance of CAZymes found in termite guts.
   :header: "sample","cazyme","clr_value"

   "A_heterognathus_BEC338","CE9",-0.08008254
   "A_heterognathus_BEC338","GH1",-0.08008254
   "A_heterognathus_BEC338","GH10",-0.08008254
   "A_heterognathus_BEC338","GH115",-0.08008254
   "A_heterognathus_BEC338","GH116",-0.08008254
   "A_heterognathus_BEC338","GH127",-0.08008254
   "A_heterognathus_BEC338","GH13_14",-0.08008254
   "A_heterognathus_BEC338","GH13_2",-0.08008254
   "A_heterognathus_BEC338","GH13_20",1.34193247
   "A_heterognathus_BEC338","GH13_31",-0.08008254
   "A_heterognathus_BEC338","GH13_39",-0.08008254
   "...","...",...,"...","..."

Figures
^^^^^^^

Figures are given in **PDF** and **HTML** format and are publication-ready. They can be found in the :file:`{{OUTDIR}}/figures` directory.

.. thumbnail:: _static/MAG_metabolic_pathways.png

**Example PDF** Presence/absence of metabolic pathway genes involved in lignocellulose digestion found in MAGs (metagenome-assembled genomes). Completeness and contamination of MAGs is given on the left site; lighter colours indicate higher completeness and less contamination, respectively. Purple squares indicate absence of genes and yellow squares indicate presence of genes. Note that some genes are shared between pathways, for instance genes of the Rnf complex.

Download here an :download:`example HTML file <_static/rel_abundance_of_bacteria_and_archaea_in_metagenomes.html>`
